ledger/src/blockstore.rs

```diff
@@ -1930,11 +1930,17 @@ impl Blockstore {
 
     /// The first complete block that is available in the Blockstore ledger
     pub fn get_first_available_block(&self) -> Result<Slot> {
-        let mut root_iterator = self.rooted_slot_iterator(self.lowest_slot())?;
-        // The block at root-index 0 cannot be complete, because it is missing its parent
-        // blockhash. A parent blockhash must be calculated from the entries of the previous block.
-        // Therefore, the first available complete block is that at root-index 1.
-        Ok(root_iterator.nth(1).unwrap_or_default())
+        let mut root_iterator = self.rooted_slot_iterator(self.lowest_slot_with_genesis())?;
+        let first_root = root_iterator.next().unwrap_or_default();
+        // If the first root is slot 0, it is genesis. Genesis is always complete, so it is correct
+        // to return it as first-available.
+        if first_root == 0 {
```

Comment:
will we hit this when restarting a node from snapshot and empty rocksdb dir? @joeaba had reported BT upload breakage when restarting warehouse nodes in this configuration as first available was reporting zero (and last the snapshot slot IIRC).  I figured it was the original `unwrap_or_default()` which makes it impossible to discern between starting from genesis and simply having no ledger

Reply:
In my testing on a node booting from snapshot and empty ledger, once the RPC service was available the first_root was set to the snapshot slot. This was true both before and after this PR.

If there were no other roots, first-available would report 0, though, because we are looking at the index-1 item in the iterator. I'm guessing this was what happened in the case you're referring to?
As you say, this was and is due to the `unwrap_or_default()`. Should we return an error instead?

Reply:
Something like "node still booting"...?

Reply:
I've been trying to decide what the least wrong thing to do is here.  Error seems right, but I'm not sure what the downstream effects might be.  We can punt to another PR though I think

Reply:
> We can punt to another PR though I think

Sg. Possibly we also need to look at error-handling in the bigtable uploader as well